tools/fork-network/src/cli.rs

```diff
@@ -237,30 +237,47 @@ impl ForkNetworkCommand {
         let head = store.get_ser::<Tip>(DBCol::BlockMisc, FINAL_HEAD_KEY)?.unwrap();
         let shard_layout = epoch_manager.get_shard_layout(&head.epoch_id)?;
         let all_shard_uids = shard_layout.get_shard_uids();
-        let fork_head = get_fork_head(&all_shard_uids, store.clone())?;
-        tracing::info!(?fork_head);
+        let num_shards = all_shard_uids.len();
+        // Flat state can be at different heights for different shards.
+        // That is fine, we'll simply lookup state root for each .
+        let fork_heads = get_fork_heads(&all_shard_uids, store.clone())?;
+        tracing::info!(?fork_heads);
 
         let chain =
             ChainStore::new(store.clone(), near_config.genesis.config.genesis_height, false);
 
-        let header = chain.get_block_header(&fork_head.hash)?;
-        let epoch_id = header.epoch_id();
-        let num_shards = shard_layout.num_shards();
-        let block_hash = fork_head.hash;
-        let block_height = fork_head.height + 1;
+        // Move flat storage to the max height for consistency across shards.
+        let (block_height, desired_block_hash) =
+            fork_heads.iter().map(|head| (head.height, head.hash)).max().unwrap();
 
+        let desired_block_header = chain.get_block_header(&desired_block_hash)?;
+        let epoch_id = desired_block_header.epoch_id();
+        let flat_storage_manager = FlatStorageManager::new(store.clone());
+
+        // Advance flat heads to the same (max) block height to ensure
+        // consistency of state across the shards.
         let state_roots: Vec<StateRoot> = (0..num_shards)
             .map(|shard_id| {
-                let shard_uid = epoch_manager.shard_id_to_uid(shard_id, &epoch_id).unwrap();
-                let chunk_extra = chain.get_chunk_extra(&fork_head.hash, &shard_uid).unwrap();
-                *chunk_extra.state_root()
+                let epoch_id = &epoch_ids[shard_id];
+                let shard_uid =
+                    epoch_manager.shard_id_to_uid(shard_id as ShardId, epoch_id).unwrap();
+                flat_storage_manager.create_flat_storage_for_shard(shard_uid).unwrap();
+                let flat_storage =
+                    flat_storage_manager.get_flat_storage_for_shard(shard_uid).unwrap();
+                flat_storage.update_flat_head(&desired_block_hash, true).unwrap();
+                let chunk_extra = chain.get_chunk_extra(&desired_block_hash, &shard_uid).unwrap();
+                let state_root = chunk_extra.state_root();
+                tracing::info!(?shard_id, ?epoch_id, ?state_root);
```

Comment:
nit: can you add a message to the log? Something like "Moved the flat state head to common height"